Optimizations & integration of alternate algorithms into pipeline

LachlanStuart opened the pull request:
This PR adds:
* The optimization to pre-select 20 decoy adducts during DB creation, instead of annotating all 80 adducts every time
* The optimization to put DS segments into a DataFrame so that more datatype can be tuned per-column to reduce the size by 33%
* Lots of extra validation steps. If you run `pipeline(debug_validation=True)` it will do some checks on the temp data after many of the steps.
* Assertions to check when [CloudObjects are created with duplicate keys](https://github.com/pywren/pywren-ibm-cloud/issues/322). Even with the smallest DB and DS, this seemed to happen multiple times per full pipeline run. In these cases it's better to restart the step than to continue with bad data that could cause OOMs / bad results later.
* Integration of `build_database` and `calculate_centroids` into `Pipeline`, including using `CloudObject`s, so that it's easier to switch between the two implementations of `build_database`
* Integration of the "big VM-style" versions of DB building and DS segmentation into the main pipeline. The old codepath can be run with `Pipeline(vm_algorithm=False)`. Note that the temp data differs between the two implementations, so modes shouldn't be mixed in the same pipeline.
* Some enhancements to the `PipelineCacher` so that data that is both DS- and DB-specific isn't accidentally shared between runs with the same DS but different DBs.

I've intentionally kept the old codepath intact in case you guys plan to compare before & after the VM conversion.

I have tested it successfully with `pipeline(vm_algorithm=True, debug_validation=True)` with ds1/db1. I also tested the first optimization with ds2/db2 and successfully ran `check_results`, however my internet has been to slow to do that with the second optimzation, since that optimization requires downloading & re-uploading the full dataset every run.

I haven't been able to finish a test with the old code (`Pipeline(vm_algorithm=False)`) because of several PyWren issues: 
* The duplicate CloudObject keys bug
* For some reason the invoker often hangs at exactly 88/256 invocations when running `segment_centr_chunk`. I haven't been able to find out why. All I can see is that PyWren hasn't even started the remaining invocations yet. EDIT: [I found the cause](https://github.com/pywren/pywren-ibm-cloud/issues/324)

LachlanStuart committed:
Handle prefix inside PipelineCacher

LachlanStuart committed:
Move molecular_db into the pipeline and use cobjects

LachlanStuart committed:
Move molecular_db_local into the pipeline and use cobjects

LachlanStuart committed:
Make cache depend on DS and/or DB based on whether the depends on the DS and/or DB config

LachlanStuart committed:
Integrate pre-annotation decoy adduct selection optimization

LachlanStuart committed:
Add validation functions for build_database and segment_centroids

LachlanStuart committed:
More validation & cleanups

LachlanStuart committed:
Copy VM-based DS segmentation into annotation_pipeline

LachlanStuart committed:
Integrate segment_ds_vm.py
Update ds_segms to use 32/64/32-bit DataFrame

LachlanStuart committed:
Fix reading ds segments with old codepath

LachlanStuart committed:
Fix reading DS segments
Make vm_algorithm a constructor argument

omerb01 commented on annotation_pipeline/segment_ds_vm.py:
```diff
@@ -0,0 +1,206 @@
+    return imzml_size, ibd_size
+
+
+def ds_imzml_path(ds_data_path):
+    return next(str(p) for p in Path(ds_data_path).iterdir()
+                if str(p).lower().endswith('.imzml'))
```
we already got this function in `utils.py`

omerb01 reviewed (changes requested):
@LachlanStuart impressive, will be very convenient for testing and comparing.
these are my comments (optional changes can wait for next PRs):

**mandatory changes request**:
- use a private connection to dataset storage, instead of using a public url in the part of dataset segmentation inside a VM. it will download the dataset much more faster.
- add "storage_backend" config variable in the config template.

**optional changes request**:
- optimize dataset segmentation script - currently the code doesn't utilise the available resources good enough. the ideal approach is to use memory bounded parallelism jobs sharing the same local storage (inside the VM) to chunk and segment the data - by this way, the more resources you have, it will dynamically tune to utilise them all.
- change all local executions prints, that are intended to run on a VM, so it will use the "logger" and the logs will be consistent with the other stages.

LachlanStuart committed:
Update segment_ds_vm to use logger instead of print

LachlanStuart committed:
Move datasets back into COS

LachlanStuart committed:
Add storage_backend to config template

LachlanStuart committed:
Do dataset segmentation in parallel
Don't re-sort DS segments after merging, because they're already sorted

LachlanStuart committed:
Make DS segment filenames consistent

LachlanStuart commented:
@omerb01 I've made all requested changes.

I've used `"cos://"` strings for the dataset paths and split them into bucket & key when they're used, because I really don't want to go back to passing around bucket names separately. I'm not sure if there's a cleaner solution available...

The `storage_backend` field in config is hopefully only temporary - [I'm hoping soon a Storage instance will be accessible via `pw.storage`](https://github.com/pywren/pywren-ibm-cloud/issues/320), instead of needing to manually create it. But you're right, that field should be in the config template if it's needed.

`segment_ds_vm` now writes `n_chunks * n_segments` files and then merges them, instead of appending to existing files. Appending to existing files wasn't a good strategy for multi-threaded writes. Hopefully because this is disk I/O we won't see the same bottlenecks that required you to implement a two-stage segmentation.

I've also removed the code that sorts by `mz` in `read_ds_segments` because both the Function and VM implementations now save sorted DS segments. As long as the segments are concatenated in ascending `segm_i` order, the result should also be sorted. It's less work (and there's less memory pressure) to sort a segment once when it's written, than to sort it every time it's read.

omerb01 reviewed (changes requested):
@LachlanStuart thanks for the changes!

**mandatory changes requests**:
- make sure that the cached data of a pipeline execution that run on a VM doesn't collide with the cached data of a total serverless pipeline execution. it seems that when we use cache in both scenarios and switch between the two states, it sometimes uses the cache and sometimes it doesn't.
For example, when we use serverless algorithm and then switch to vm algorithm, it rebuilds the database but uses the same peaks from before and then apply the vm dataset segmentation but also uses the same cached annotations and db segments from before.
I suggest to use a different "storage directory" for each state (VM or not).

**optional changes requests**:
- add logger prints of loaded cached data for the new parts that use cache: `build_database()`, `calculate_peaks()` and segmentation over VM. I mean for "X loaded from cahce" logs
- minor thing - when downloading the dataset in the VM algorithm, please add a logger print "download dataset..." so we will know that it isn't stuck.

after handle these change requests, you can merge the PR, Thanks!

LachlanStuart committed:
Move cache into namespaces so that vm/non-vm caches are completely separate

LachlanStuart committed:
Extra logging per PR feedback
